firebase-firestore/src/test/java/com/google/firebase/firestore/core/FilterTest.java

```diff
@@ -0,0 +1,322 @@
+    // (A | B) & (C | D) --> (A & C) | (A & D) | (B & C) | (B & D)
+    CompositeFilter expectedResult =
+        orFilters(andFilters(A, C), andFilters(A, D), andFilters(B, C), andFilters(B, D));
+    assertThat(applyDistribution(orFilters(A, B), orFilters(C, D))).isEqualTo(expectedResult);
+  }
+
+  @Test
+  public void testFieldFilterComputeDnf() {
+    assertThat(computeDnf(A)).isEqualTo(A);
+    assertThat(DnfTransform(andFilters(A))).isEqualTo(Collections.singletonList(A));
+    assertThat(DnfTransform(orFilters(A))).isEqualTo(Collections.singletonList(A));
+  }
+
+  @Test
+  public void testComputeDnfFlatAndFilter() {
+    CompositeFilter compositeFilter = andFilters(A, B, C);
+    assertThat(computeDnf(compositeFilter)).isEqualTo(compositeFilter);
+    assertThat(DnfTransform(compositeFilter)).isEqualTo(Collections.singletonList(compositeFilter));
+  }
+
+  @Test
+  public void testComputeDnfFlatOrFilter() {
+    CompositeFilter compositeFilter = orFilters(A, B, C);
+    assertThat(computeDnf(compositeFilter)).isEqualTo(compositeFilter);
+    Filter[] expectedDnfTerms = {A, B, C};
+    assertThat(DnfTransform(compositeFilter)).isEqualTo(Arrays.asList(expectedDnfTerms));
+  }
+
+  @Test
+  public void testComputeDnf1() {
+    // A & (B | C) --> (A & B) | (A & C)
+    CompositeFilter compositeFilter = andFilters(A, orFilters(B, C));
+    Filter[] expectedDnfTerms = {andFilters(A, B), andFilters(A, C)};
+    CompositeFilter expectedResult = orFilters(expectedDnfTerms);
+    assertThat(computeDnf(compositeFilter)).isEqualTo(expectedResult);
+    assertThat(DnfTransform(compositeFilter)).isEqualTo(Arrays.asList(expectedDnfTerms));
+  }
+
+  @Test
+  public void testComputeDnf2() {
+    // ((A)) & (B & C) --> A & B & C
+    CompositeFilter compositeFilter = andFilters(andFilters(andFilters(A)), andFilters(B, C));
+    CompositeFilter expectedResult = andFilters(A, B, C);
+    assertThat(computeDnf(compositeFilter)).isEqualTo(expectedResult);
+    assertThat(DnfTransform(compositeFilter)).isEqualTo(Collections.singletonList(expectedResult));
+  }
+
+  @Test
+  public void testComputeDnf3() {
+    // A | (B & C)
+    CompositeFilter compositeFilter = orFilters(A, andFilters(B, C));
+    assertThat(computeDnf(compositeFilter)).isEqualTo(compositeFilter);
+    Filter[] expectedDnfTerms = {A, andFilters(B, C)};
+    assertThat(DnfTransform(compositeFilter)).isEqualTo(Arrays.asList(expectedDnfTerms));
+  }
+
+  @Test
+  public void testComputeDnf4() {
+    // A | (B & C) | ( ((D)) | (E | F) | (G & H) ) --> A | (B & C) | D | E | F | (G & H)
+    CompositeFilter compositeFilter =
+        orFilters(
+            A,
+            andFilters(B, C),
+            orFilters(andFilters(orFilters(D)), orFilters(E, F), andFilters(G, H)));
+    Filter[] expectedDnfTerms = {A, andFilters(B, C), D, E, F, andFilters(G, H)};
+    CompositeFilter expectedResult = orFilters(expectedDnfTerms);
+    assertThat(computeDnf(compositeFilter)).isEqualTo(expectedResult);
```

Comment:
In general, we should only be testing public APIs. This makes it harder to test the logic in complicated helper methods, so it the test cases above it makes sense that you directly invoke `applyX()`. I wonder though if it makes sense in the tests here as you are invoking both the public and non-public API.